### ISSUE TITLE
Update whitelisting does not correctly whitelist dependencies of new packages to be installed

### ISSUE THREAD

**Issue by naderman** [composer/composer#5438]
composer.json.1:

``` json
{
    "repositories": [
        {
            "packagist": false
        },
        {
            "type": "package",
            "package": {
                "name": "a/a",
                "version": "1.0.0",
                "type": "metapackage",
                "require": {
                    "a/b": "*"
                }
            }
        },
        {
            "type": "package",
            "package": {
                "name": "a/b",
                "version": "1.0.0",
                "type": "metapackage"
            }
        },
        {
            "type": "package",
            "package": {
                "name": "a/c",
                "version": "1.0.0",
                "type": "metapackage",
                "require": {
                    "a/b": "2.0.0"
                }
            }
        }
    ],
    "require": {
        "a/a": "1.0.0"
    }
}
```

composer.json.2:

``` json
{
    "repositories": [
        {
            "packagist": false
        },
        {
            "type": "package",
            "package": {
                "name": "a/a",
                "version": "1.0.0",
                "type": "metapackage",
                "require": {
                    "a/b": "*"
                }
            }
        },
        {
            "type": "package",
            "package": {
                "name": "a/b",
                "version": "1.0.0",
                "type": "metapackage"
            }
        },
        {
            "type": "package",
            "package": {
                "name": "a/b",
                "version": "2.0.0",
                "type": "metapackage"
            }
        },
        {
            "type": "package",
            "package": {
                "name": "a/c",
                "version": "1.0.0",
                "type": "metapackage",
                "require": {
                    "a/b": "2.0.0"
                }
            }
        }
    ],
    "require": {
        "a/a": "1.0.0",
        "a/c": "1.0.0"
    }
}
```

When I run this command:

```
COMPOSER=composer.json.1 composer update
COMPOSER=composer.json.2 composer update a/c --with-dependencies --dry-run
```

I get this output:

```
Loading composer repositories with package information
Updating dependencies (including require-dev)
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Can only install one of: a/b[2.0.0, 1.0.0].
    - Can only install one of: a/b[2.0.0, 1.0.0].
    - a/c 1.0.0 requires a/b 2.0.0 -> satisfiable by a/b[2.0.0].
    - Installation request for a/c 1.0.0 -> satisfiable by a/c[1.0.0].
    - Installation request for a/b (installed at 1.0.0) -> satisfiable by a/b[1.0.0].

```

And I expected this to happen:
- a/b should be updated to 2.0
- a/c 1.0 should be installed

**Comment by naderman** [comment]
This may be rather hard to change. At the time of whitelisting only locally available packages can be considered. The new package is not one of them yet, so its dependencies are not known yet and cannot be whitelisted.

If we only lock composer.json defined dependencies (default whitelist with dep behaviour if package is present), then we may accidentally update other dependencies that are not related to the new package.

One way of fixing this would be to then roll back any unrelated dependency changes in the resulting solver transaction. Another would be to pass in the information about which packages may be updated if really needed to the solver and have it deal with this, however I have no immediate concept for how to implement that.

**Comment by slbmeh** [comment]
The rollback approach sounds flaky and almost as difficult as the optional whitelist.  Perhaps `PolicyInterface` can be leverage to prefer installed versions over newer/older from the dep tree of whitelisted packages.

**Comment by Seldaek** [comment]
With #8558 this now errors as:

```
  Problem 1
    - Root composer.json requires a/c 1.0.0 -> satisfiable by a/c[1.0.0].
    - a/c 1.0.0 requires a/b 2.0.0 -> found a/b[2.0.0] but the package is fixed to 1.0.0 (lock file version) by a partial update and that version does not match. Make sure you whitelist it for update.
```

It still errors, but at least it's quite clear that a/b 2.0 is required, is found, and that a/b needs to be whitelisted.

I would say that's an actionable enough error message that we could close this, given that fixing the root cause is very much non-trivial. What do you all think?

**Comment by naderman** [comment]
That's nice, but I think the issue remains, so I wouldn't close this yet.

**Comment by Seldaek** [comment]
Fixed by #8717
